server/src-lib/Hasura/Server/Auth/WebHook.hs

```diff
@@ -0,0 +1,150 @@
+module Hasura.Server.Auth.WebHook
+  ( AuthHookType(..)
+  , AuthHookG (..)
+  , AuthHook
+  , userInfoFromAuthHook
+  ) where
+
+import           Control.Exception         (try)
+import           Control.Lens
+import           Control.Monad.Trans.Maybe
+import           Data.Aeson
+import           Data.Time.Clock           (UTCTime, addUTCTime, getCurrentTime)
+import           Hasura.Server.Version     (HasVersion)
+
+import qualified Data.Aeson                as J
+import qualified Data.ByteString.Lazy      as BL
+import qualified Data.HashMap.Strict       as Map
+import qualified Data.Text                 as T
+import qualified Network.HTTP.Client       as H
+import qualified Network.HTTP.Types        as N
+import qualified Network.Wreq              as Wreq
+
+import           Data.Parser.CacheControl
+import           Data.Parser.Expires
+import           Hasura.HTTP
+import           Hasura.Logging
+import           Hasura.Prelude
+import           Hasura.RQL.Types
+import           Hasura.Server.Logging
+import           Hasura.Server.Utils
+
+
+data AuthHookType
+  = AHTGet
+  | AHTPost
+  deriving (Eq)
+
+instance Show AuthHookType where
+  show AHTGet  = "GET"
+  show AHTPost = "POST"
+
+isPost :: AuthHookType -> Bool
+isPost AHTGet  = False
+isPost AHTPost = True
+
+
+data AuthHookG a b
+  = AuthHookG
+  { ahUrl  :: !a
+  , ahType :: !b
+  } deriving (Show, Eq)
+
+type AuthHook = AuthHookG T.Text AuthHookType
+
+withAuthHook :: AuthHook -> (T.Text -> a) -> (T.Text -> a) -> a
+withAuthHook (AuthHookG u t) ifGet ifPost =
+  bool (ifGet u) (ifPost u) $ isPost t
+
+hookMethod :: AuthHook -> N.StdMethod
+hookMethod = bool N.GET N.POST . isPost . ahType
+
+
+-- | Extracts 'UserInfo' from the webhook answer.
+--   This uses the provided manager to make the call, extracts
+--   information such as response body and status, and delegates to a
+--   local 'mkUserInfoFromResp'.
```

Comment:
I don’t think this comment is very helpful—it mostly just describes what the implementation does. It’s a documentation comment, so it probably ought to be useful to *callers* of `userInfoFromAuthHook`, but the fact that it delegates to a local function is just an implementation detail.

If I were to write a comment at all, I’d leave it at something as simple as this:

> Makes an authentication request to the given `AuthHook` and returns `UserInfo` parsed from the response, plus an expiration time if one was returned.

Mentioning that the second result is an expiration time is the most useful bit, since the purpose of the other arguments/results can be inferred from their types.

Reply:
Done!